### PR TITLE
 Use argparse for entry point

### PR DESCRIPTION
**Pull request by satreix**
Use python argparse module instead of custom code.

### CODE REVIEW

**Comment by SickRage2** [comment]
Thanks for the pull request! Before a real human comes by, please make sure your PR has all of the below criteria checked
- [ ] Give a description on what the PR is for.
- [x] Make sure your PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review

Please make sure you also read [contribution guide](https://github.com/SickRage/SickRage/blob/master/contributing.md) and followed all of the steps.

Thanks!

**Comment by Arcanemagus** [comment]
Nice, always good to replace custom code with a library if possible.

**Comment by OmgImAlexis** [comment]
Should the comments not start with a capital like the old code did?

**Comment by satreix** [comment]
@OmgImAlexis I guess no, default argparse `--help` start without one. That gives us the following :question: text:

```
usage: SickBeard.py [-h] [-q] [--nolaunch] [-d] [--pidfile PIDFILE] [-p PORT]
                    [--datadir DATADIR] [--config CONFIG] [--noresize]

optional arguments:
  -h, --help            show this help message and exit
  -q, --quiet           disables logging to console
  --nolaunch            suppress launching web browser on startup
  -d, --daemon          run as double forked daemon (includes options --quiet
                        --nolaunch)
  --pidfile PIDFILE     combined with --daemon creates a pidfile (full path
                        including filename)
  -p PORT, --port PORT  override port to listen on
  --datadir DATADIR     override folder (full path) as location for storing
                        database, configfile, cache and logfiles
  --config CONFIG       override config filename (full path) to load
                        configuration from
  --noresize            prevent resizing of the banner/posters even if PIL is
                        installed
```

Is it fine with you?

File: SickBeard.py [hunk 1]
```diff
@@ -23,35 +23,18 @@
...
-  -p,  --port=[PORT]     Override default/configured port to listen on
-       --datadir=[PATH]  Override folder (full path) as location for
-                         storing database, config file, cache, and log files
-                         Default SickRage directory
-       --config=[FILE]   Override config filename for loading configuration
-                         Default config.ini in SickRage directory or
-                         location specified with --datadir
-       --noresize        Prevent resizing of the banner/posters even if PIL
-                         is installed
+SickBear
```
Comment (miigotu) [inline SickBeard.py]: we dont want any sick bears, it might be rabies

Comment (satreix) [inline SickBeard.py]: My bad

**Comment by OmgImAlexis** [comment]
@satreix all's good then, just thought I'd mention it.

**Comment by labrys** [comment]
@satreix For the formatting alignment you could subclass argparse.HelpFormatter.

**Comment by satreix** [comment]
@labrys do we really need that extra complexity?

For now what I have is:

```
usage: SickBeard.py [-h] [-q] [--nolaunch] [-d] [--pidfile PIDFILE] [-p PORT]
                    [--datadir DATADIR] [--config CONFIG] [--noresize]

Automatic Video Library Manager for TV Shows. It watches for new episodes of
your favorite shows, and when they are posted it does its magic.

optional arguments:
  -h, --help            show this help message and exit
  -q, --quiet           disables logging to console
  --nolaunch            suppress launching web browser on startup
  -d, --daemon          run as daemon (includes options --quiet --nolaunch)
  --pidfile PIDFILE     combined with --daemon creates a pidfile (full path)
  -p PORT, --port PORT  port to listen on
  --datadir DATADIR     folder full path to store database, configfile, cache
                        and logfiles
  --config CONFIG       config file full path to load configuration from
  --noresize            prevent resizing of the banner/posters even if PIL is
                        installed
```

80 columns, complete. That is sufficient don't you think?

**Comment by satreix** [comment]
I also split the arguments parsing away from the `.start()` method to allow _SickRage_ to be started without a _tty_. This introduces the `.setup_from_commend_line()` method.

**Comment by labrys** [comment]
@satreix argparse. HelpFormatter subclasses are generally very short so I think the more professional look would be worth the effort, however I wouldn't let that hold up the PR just an improvement to think about for the future. e.g. this is the code for all three subclasses of the formatter:

``` python
class RawDescriptionHelpFormatter(HelpFormatter):
    """Help message formatter which retains any formatting in descriptions.

    Only the name of this class is considered a public API. All the methods
    provided by the class are considered an implementation detail.
    """

    def _fill_text(self, text, width, indent):
        return ''.join([indent + line for line in text.splitlines(True)])


class RawTextHelpFormatter(RawDescriptionHelpFormatter):
    """Help message formatter which retains formatting of all help text.

    Only the name of this class is considered a public API. All the methods
    provided by the class are considered an implementation detail.
    """

    def _split_lines(self, text, width):
        return text.splitlines()


class ArgumentDefaultsHelpFormatter(HelpFormatter):
    """Help message formatter which adds default values to argument help.

    Only the name of this class is considered a public API. All the methods
    provided by the class are considered an implementation detail.
    """

    def _get_help_string(self, action):
        help = action.help
        if '%(default)' not in action.help:
            if action.default is not SUPPRESS:
                defaulting_nargs = [OPTIONAL, ZERO_OR_MORE]
                if action.option_strings or action.nargs in defaulting_nargs:
                    help += ' (default: %(default)s)'
        return help
```

**Comment by labrys** [comment]
@satreix I'll try to get around to looking at this and testing it tonight.  @neoatomic are any of the options used on Syno?

**Comment by medariox** [comment]
@satreix 
While the idea behind this PR is certainly good, we still have to make sure that everything runs as expected. I tested some arguments and encountered multiple issues.

Off the top of my head, I remember these few:
- argument `-q` is not quiet
- `ctrl+c` doesn't terminate the process
- restarting SR within the UI spawns a new process and leaves a zombie process behind that never terminates
- argument `-d` doesn't include argument `-q`, although specified in the arguments explanation

Please make sure you solve the above issues and test the PR thoroughly, otherwise we won't be able to merge it. The tests above have been run with Ubuntu 15.04 as non-root user.
Thank you!

**Comment by labrys** [comment]
@satreix also consider adding an option to do backups/restores from the command line.. would be nice i think.

**Comment by labrys** [comment]
@satreix let me know if you need any help addressing the issues @medariox mentioned.  The addition of backup/restore i mentioned could be a future pr.

**Comment by miigotu** [comment]
I will also be adding an arg to completely disable and hide the updater, for packaging reasons on a dist that wants to distribute SR.

**Comment by satreix** [comment]
@medariox wrote:

> argument -q is not quiet

According to the newly added tests, this not the argument parsing fault: https://github.com/SickRage/SickRage/blob/2baace9361451c0889e8ddeb5607ec820c9fa761/tests/cli/argument_parsing.py#L14

Here is this tests output:

```
--config CONFIG: config file full path to load configuration from ... passed
-d, --daemon: run as daemon (includes options --quiet --nolaunch) ... passed
-d, --daemon: run as daemon (includes options --quiet --nolaunch) ... skipped
--datadir DATADIR: folder full path to store database, configfile, cache ... passed
--noresize: prevent resizing of the banner/posters even if PIL is ... passed
--nolaunch: suppress launching web browser on startup ... passed
--pidfile PIDFILE: combined with --daemon creates a pidfile (full path) ... passed
-p PORT, --port PORT: port to listen on ... passed
-q, --quiet: disables logging to console ... passed
```

> ctrl+c doesn't terminate the process

What is the code responsible for this behavior please? :smile: 

> restarting SR within the UI spawns a new process and leaves a zombie process behind that never terminates

Same here, where is the code? :smile: 

> argument -d doesn't include argument -q, although specified in the arguments explanation

Not according to this (that passes on archlinux): https://github.com/SickRage/SickRage/blob/2baace9361451c0889e8ddeb5607ec820c9fa761/tests/cli/argument_parsing.py#L46

Thank you for testing.

**Comment by miigotu** [comment]
Just to make sure that no one merges this before I can make a detailed review when it is done.

**Comment by OmgImAlexis** [comment]
@satreix ctrl+c sends the`intr` command. Have a look [here](https://stackoverflow.com/questions/1112343/how-do-i-capture-sigint-in-python).

**Comment by fernandog** [comment]
@satreix hey, still working on this? thanks!

**Comment by satreix** [comment]
I am still testing the reported bugs.

**Comment by satreix** [comment]
@medariox, after some debugging:

#### Quiet

> argument -q is not quiet

What is still showing?

> argument -d doesn't include argument -q, although specified in the arguments explanation

It does, I guess this is related to the previous one.

#### Process termination

> `ctrl+c` doesn't terminate the process

It gets stuck in https://github.com/SickRage/SickRage/blob/master/SickBeard.py#L380. Trying to figure it out now.

> restarting SR within the UI spawns a new process and leaves a zombie process behind that never terminates

I guess this is related to the previous one.

**Comment by labrys** [comment]
@satreix It's supposed to loop at L380, that's the main loop for SR.

**Comment by satreix** [comment]
@labrys even on `ctrl+c`?

**Comment by labrys** [comment]
@satreix No but that line isn't catching anything either so it wouldn't be whats keeping ctl-c from working.

File: SickBeard.py [hunk 2]
```diff
@@ -544,11 +520,11 @@ def shutdown(self, event):
                     logger.shutdown()  # pylint: disable=no-member
                     subprocess.Popen(popen_list, cwd=os.getcwd())
 
-        # Make sure the logger has stopped, just in case
+        # Make sure the logger has stopped
         logger.shutdown()  # pylint: disable=no-member
-        os._exit(0)  # pylint: disable=protected-access
```
Comment (miigotu) [inline SickBeard.py:549]: @satreix here.

**Pull request closed without merging**